linkerd/app/inbound/src/lib.rs

```diff
@@ -217,86 +220,93 @@ where
+                let detect_timeout = cfg.proxy.detect_protocol_timeout;
+                let require_id = cfg.require_identity_for_inbound_ports.clone();
+
+                http.push_map_target(HttpAccept::from)
+                    .push(svc::UnwrapOr::layer(
+                        // When HTTP detection fails, forward the connection to the application as
+                        // an opaque TCP stream.
+                        tcp.into_stack()
+                            .push_map_target(TcpEndpoint::from)
+                            .push_on_response(svc::BoxService::layer())
+                            .into_inner(),
+                    ))
+                    .push_map_target(detect::allow_timeout)
+                    .push(svc::BoxNewService::layer())
+                    .push(detect::NewDetectService::layer(
+                        detect_timeout,
+                        http::DetectHttp::default(),
+                    ))
+                    .push_request_filter(require_id)
                     .push(rt.metrics.transport.layer_accept())
-                    .check_new_service::<TcpAccept, _>()
-                    .instrument(|_: &TcpAccept| debug_span!("forward"))
-                    .into_inner(),
-            )
-            .check_new_service::<T, I>()
-            .push_on_response(svc::BoxService::layer())
-            .push(svc::BoxNewService::layer())
-            .push_switch(
-                PreventLoop::from(server_port).to_switch(),
-                self.push_tcp_forward(server_port)
-                    .push_direct(gateway)
-                    .stack
-                    .instrument(|_: &_| debug_span!("direct"))
-                    .into_inner(),
-            )
-            .instrument(|a: &T| {
-                let OrigDstAddr(target_addr) = a.param();
-                info_span!("server", port = target_addr.port())
+                    .push_request_filter(TcpAccept::try_from)
+                    .push(svc::BoxNewService::layer())
+                    .push(tls::NewDetectTls::layer(
+                        rt.identity.clone(),
+                        detect_timeout,
+                    ))
+            })
+            .map_stack(|cfg, _, detect| {
+                let disable_detect = cfg.disable_protocol_detection_for_ports.clone();
+                detect
+                    .instrument(|_: &_| debug_span!("proxy"))
+                    .push_switch(
+                        move |t: T| {
+                            let OrigDstAddr(addr) = t.param();
+                            if !disable_detect.contains(&addr.port()) {
+                                Ok::<_, Never>(svc::Either::A(t))
+                            } else {
+                                Ok(svc::Either::B(TcpAccept::port_skipped(t)))
+                            }
+                        },
+                        opaque
+                            .instrument(|_: &TcpAccept| debug_span!("forward"))
```

Comment:
very unimportant nit: it seems a little inconsistent that we add the span to `opaque` here, but the span in the direct stack is added in the let binding above. very minor preference for making these consistent, but this absolutely does not actually matter at all :)

Reply:
agreed, will sort this out with the followups.